[Kartikeya] iP

kxrt opened the pull request:
<h1 align="center">Artemis 🚀</h1>

> "I like this" - Me ([source](https://github.com/nus-cs2103-AY2223S1/ip/pull/20))

## History 

Earlier versions of this chatbot were named *Apollo*, in honor of the first manned missions to the Moon. To honor NASA's latest venture to the Moon, *Apollo* was renamed to *Artemis* for the latest version simultaneously.

## General

*Artemis* is a chatbot that eases your mental load, keeping track of your:
- ToDos
- Events
- Deadlines

*Artemis* is very simple to use! It is:
- text-based
- easy to learn
- ~~FAST~~ *SUPER FAST* to use

The best part? *Artemis* is <ins>COMPLETELY FREE</ins> to use! 🤩

## Setup

Here's a step-by-step process on how to set *Artemis* up:
1. Download the latest release's `duke.jar` file from [here](https://github.com/kxrt/ip/releases).
2. Navigate to where the file is stored on your CLI and run with `java -jar duke.jar`.
3. Add your tasks, and instantly become more productive!

**Note:** It's recommended to store the `jar` file in a separate folder for *Artemis*.

## Features

Here's *Artemis*' feature list:
- [x] Managing ToDos
- [x] Managing Deadlines
- [x] Managing Events
- [x] Keyword finding *(More improvements incoming!)*
- [ ] GUI *(Coming soon!)*
- [ ] Reminders *(Coming soon!)*

## Available Commands

| Command | Description |
| -- | -- |
| todo `desc` | Creates a todo with the given description.
| deadline `desc` /by `time` | Creates a deadline with the given description set by the given time. |
| event `desc` /at `time` | Creates an event with the given description set at the given time. |
| list | Lists all currently saved tasks. |
| mark `num` | Marks task `num` from `list` as complete. |
| unmark `num` | Marks task `num` from `list` as incomplete. |
| delete `num` | Deletes task `num` from `list` permanently. |
| find `text` | Returns all tasks whose descriptions match the exact text. |
| clear | Clears window of all messages. |
| help | Shows all available commands. |
| bye | Saves all tasks to storage and exits Artemis. |

**Note:** Time input **must** be formatted in a 24-hour format as: `YYYY-MM-DD HHHH` 

---
*Artemis* is written in Java 11. Here's the `main` method:
```java
public static void main(String[] args) {
    new Duke().run();
}
```

Commit:
Add Gradle support

Commit:
build.gradle: Update version to 8.29

kxrt committed:
Level-1

- Echo commands that are not "bye"
- Exit if "bye" command is issued

kxrt committed:
Level-2

- If "list" command is issued, print list of stored items
- Otherwise, add input string to stored item list

kxrt committed:
Level-3

- Mark and unmark tasks as done or not done
- Allow switching tasks between done/not done

kxrt committed:
Store DukeTasks in Collection instead of array

kxrt committed:
Level-4

- Add ToDo, Deadline, Event tasks
- Update application styling

kxrt committed:
A-TextUiTesting

- Set up I/O redirection testing
- Clarified input and output constraints

kxrt committed:
Level-5

- Create DukeException exception class
- Throw DukeException on incorrect inputs

kxrt committed:
Level-6

- Remove tasks when "delete" command is issued
- Colorise output strings for styling

kxrt committed:
Level-6

- Remove tasks when "delete" command is issued
- Colorise output strings for styling
- Check for missing descriptions

kxrt committed:
Level-6

kxrt committed:
Level-6

- Catch more incorrect arguments as exceptions
- Add javadocs

kxrt committed:
Level-6

- Change styling
- Change access modifiers for better encapsulation

kxrt committed:
Implement saving task data to hard disk

Level-7
- Rename DukeList to TaskList for consistency
- Use file I/O operations to save task data to hard disk
- Create data/duke.txt if it doesn't exist on the user's device
- Implement a Storage class to encapsulate the above

kxrt committed:
Implement formatting of date/time inputs

Level-8
- Create DukeDateTimeFormatter class with a static method to encapsulate formatting in one place
- Implement DukeDateTimeFormatter in Deadline and Event classes
- Create formattedAt/formattedBy strings to represent formatted date-time strings, replace into toString

kxrt committed:
Level-8 Final

kxrt committed:
Add exception handling for incorrect date/time formats

kxrt committed:
Create abstractions to further implement OOP

- Create Ui and Parser classes and combine them with TaskList and Storage to extract out closely related code as classes
- Refactor Duke to implement these classes
- Shift all user interactions, including I/O from the console to the Ui class
- Shift all input parsing to the Parsing class

kxrt committed:
Create Command abstractions to further implement OOP

Create the following:
- ExitCommand
- ListItemsCommand
- MarkCommand
- UnmarkCommand
- DeleteCommand
- AddItemCommand

kxrt committed:
Divide classes into packages

Create the following packages:
- duke
- duke.command
- duke.exception
- duke.parser
- duke.storage
- duke.task
- duke.tasklist
- duke.ui

kxrt committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

kxrt committed:
Update Gradle configuration

kxrt committed:
Add JUnit tests for DukeException and duke.task

- Test exception getMessage method and exception type for DukeException
- Test toString and getStorageString methods in all duke.task classes

kxrt committed:
Add JavaDocs to public and non-trivial private methods

kxrt committed:
Make code formatting comply with se-edu's Java standard

kxrt committed:
Fix indentation

kxrt committed:
Implement finding tasks with keywords

Keywords are compared with descriptions of all tasks, and hence
the keywords are required to be absolutely accurate for successful
finds.

kxrt committed:
Merge branch 'branch-A-JavaDoc'

kxrt committed:
Fix merge conflicts from branch-A-CodingStandard

kxrt committed:
Merge branch 'branch-Level-9'

kxrt committed:
Implement checkstyle and fix code style

- Set indentation level to 4 spaces, and 8 spaces for wrapped lines
- Move executeAddItem and executeFindItems methods from Parser.java to AddItemCommand.java and FindCommand.java, respectively
- Add missing javadocs
- Fix javadocs formatting style

kxrt committed:
Implement checkstyle and fix code style

- Set indentation level to 4 spaces, and 8 spaces for wrapped lines
- Move executeAddItem and executeFindItems methods from Parser.java to AddItemCommand.java and FindCommand.java, respectively
- Add missing javadocs
- Fix javadocs formatting style

bensohh commented on src/main/java/duke/command/AddItemCommand.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    /**
+     * Executes item addition commands, throwing a duke.exception.DukeException on incorrect
```
Perhaps you could consider removing the repeated statement about throwing an exception here since it is already mentioned under @throws

bensohh commented on src/main/java/duke/command/AddItemCommand.java:
```diff
@@ -0,0 +1,92 @@
+     * inputs.
+     *
+     * @param newItem Parsed item addition command
+     * @return string signifying result of adding newItem to items
```
Would be good to capitalise the 'string' to 'String' just after the @return

bensohh commented on src/main/java/duke/command/AddItemCommand.java:
```diff
@@ -0,0 +1,92 @@
+     *
+     * @param newItem Parsed item addition command
+     * @return string signifying result of adding newItem to items
+     * @throws DukeException Indicates incorrect inputs
```
Maybe you could consider rephrasing it towards @throws DukeException If (something occurs)?

bensohh commented on src/main/java/duke/command/AddItemCommand.java:
```diff
@@ -0,0 +1,92 @@
+            return "";
+        }
+        DukeTask newTask;
+        switch (newItem[0]) {
+        case "todo": {
```
Good use of switch-case with correct indentation 👍

bensohh commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,24 @@
+     * Executes the command action.
+     *
+     * @param itemList list of tasks
+     * @param ui       ui interactions with user
```
Maybe removing the empty spaces between ui and ui would be neater and more standardised.

kxrt replied:
I was unsure of the proper format to follow here myself. I resorted to following the [Oracle standard](https://www.oracle.com/sg/technical-resources/articles/java/javadoc-tool.html) linked on the CS2103/T website, which mentioned: 
> Additional spaces can be inserted between the name and description so that the descriptions line up in a block.

As such, given that I've standardized it across all my code I'd prefer to stick to this, but thank you for the suggestion!

bensohh commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    /**
+     * Parses input array of strings to a string that can be used to call the find method
+     * on itemList. Calls the method and parses output to a format suitable for the ui,
+     * then returns it. Accepts an empty input after the find keyword, and returns all the
+     * tasks stored in this case. Throws a DukeException if no items were found.
```
Perhaps instead of explaining how it works, you could focus on the What (What is this method/code supposed to do?) or Why (Rationale for the implementation)?

kxrt replied:
Thank you, I've changed the Javadoc comment for this!

bensohh commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,68 @@
+     * tasks stored in this case. Throws a DukeException if no items were found.
+     *
+     * @param input array of strings input by user
+     * @return string indicating a list of tasks whose descriptions match the input string
```
As mentioned above, could change string to String after @return

bensohh commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,112 @@
+    private void createStorageFile() throws DukeException {
+        try {
+            File directory = new File(FILE_FOLDER_NAME);
+            if (!directory.exists()) {
+                directory.mkdir();
+            }
+            File saved = new File(FILEPATH);
+            if (!saved.exists()) {
+                saved.createNewFile();
+            }
```
I like the way you avoid deep nesting a formatted your code neatly based on the different possible cases, shown above as well.

bensohh reviewed (commented):
LGTM 👍 Only very minor edits to make. Overall, very neat code and good quality, avoided deep nesting which improves the readability. 😃

kxrt committed:
Update Javadocs, add GUI reqs to build.gradle

loyhongshenggg commented on src/main/java/duke/command/AddItemCommand.java:
```diff
@@ -0,0 +1,91 @@
+     * {@inheritDoc}
+     * Adds given input item to itemList, and shows the resulting message to the user.
+     */
+    @Override
```
For the header, you might want to include a @throws DukeException description

kxrt replied:
Agreed yeah, but I believe that's passed down to the method when the `inheritDoc` tag pulls from the overridden method.

loyhongshenggg commented on src/main/java/duke/command/AddItemCommand.java:
```diff
@@ -0,0 +1,91 @@
+     * Adds given input item to itemList, and shows the resulting message to the user.
+     */
+    @Override
+    public void execute(TaskList itemList, Ui ui, Storage storage) throws DukeException {
```
For the naming of method, good job on using a verb! But perhaps you might want to name it as "executeSomething"

kxrt replied:
I agree, however the problem with that is the same `execute` method is overridden by every `...Command` class, which makes it hard to find a consistent name. Perhaps executeCommand? But that seems redundant.

loyhongshenggg reviewed (commented):
Good job! Very consistent with following the coding standards!

kxrt committed:
Create GUI in JavaFX

- Additionally add "clear" command to clear GUI screen
- Style dialog boxes and overall stage design
- Create new DialogWindow, MainWindow, Launcher and Main classes for GUI purposes
- Implement FXML designs

kxrt committed:
Merge branch 'branch-Level-10'

kxrt committed:
Update build.gradle to match new classes

kxrt committed:
Make DialogBox resizable while preserving content

kxrt committed:
Allow resizing application window, update GUI

Currently,
- Text isn't wrapped in rows and extends on horizontally which
  requires users to scroll through the window to read the entire
  text
- Resizing the window isn't possible at the moment, either, and
  needs to be implemented
- Setting proper indentation depending on the dialog author will
  also enhance the user experience

Let's:
- Enable wrapping of text in Label by setting the "HBox.hgrow"
  property to "ALWAYS"
- Fix resizing issue by setting AnchorPane anchors
- Update indentation of Label text depending on who the dialog is from

Enabling wrapping with a prop is a better choice than wrapping the
Label object in a TextFlow object, preventing any side-effects.

kxrt committed:
Remove unused Ui class and text I/O testing

kxrt committed:
Add assertions

Certain assertions need to hold true for the app processes to
continue reliably. Let's add the following assertions:

- output from commands can never be empty
- itemList can never be uninitialised
- error messages should always start with "Hang on" for clarity
  that they're checked exceptions
- tasks in itemList can never be empty

kxrt committed:
Improve code quality

Code quality can be better improved, there are certain instances
of weak SLAP and code duplication leading to unnecessarily long
methods in the code.

Let's:
- Avoid code duplication in AddItemCommand.java by moving addition
  to itemList and the return statement to the bottom of the method
- Better follow SLAP in FindCommand.java by abstracting out the
  joinInput and joinOutput methods
- Avoid code duplication in Storage.java by implementing a private
  method for task checking
- Better follow SLAP in TaskList.java by implementing a method in
  DukeTask.java to check for description matches

kxrt committed:
Merge pull request #1 from kxrt/branch-A-Assertions

Add assertions

kxrt committed:
Merge branch 'branch-A-CodeQuality'

Improve code quality in the app.

kxrt committed:
Merge pull request #2 from kxrt/branch-A-CodeQuality

Improve code quality

kxrt committed:
Merge branch 'master' of https://github.com/kxrt/ip

kxrt committed:
Add Continuous Integration

Currently, unit tests need to be run manually before every push
or merge. This is tedious and requires extra steps from the
programmer, and might lead to slip ups if forgotten. Let's improve
this by setting up a CI workflow.

kxrt committed:
Merge pull request #3 from kxrt/branch-A-CI

Add Continuous Integration

kxrt committed:
Remove unnecessary docs, follow reuse policy

kxrt committed:
Remove unused imports

kxrt committed:
Implement updating of task description

- Create UpdateDescriptionCommand class for this purpose
- Add relevent exception checking regarding indexes and missing
  descriptions
- Abstract string output for IndexOutOfBoundsException to a
  single static field for better code quality

kxrt committed:
Implement updating of task time

- Create UpdateTimeCommand class for this purpose
- Add relevent exception checking regarding indexes and missing
  time
- Implement better abstraction across Parser, Deadline, DukeTask,
  Event and TaskList classes

kxrt committed:
Merge pull request #4 from kxrt/branch-C-Update

Implement updating of task description

kxrt committed:
Add "help", update identity, check input case

Users are not aware of all possible commands they can use. Also,
the identity needs to be revamped from "Apollo" to "Artemis".
Capital letter inputs are unable to be processed too.

Let's:
- Add a "help" command that lists all available commands
- Rebrand from Apollo to Artemis
- Update chat icons to match the astronaut theme
- Add new chat icon for error messages
- Allow any input case to be processed

kxrt committed:
Merge branch 'A-Personality'

kxrt committed:
Update user guide, fix styling

kxrt committed:
Create 'help' window

The list of available commands is huge and takes up a sizeable
amount of space. Let's export this to a new window, and implement
some more minor changes

- Using 'help' now opens up a new window with a list of available
  commands
- Added new icons for success and error messages
- Update user icons
- Update user guide
- Catch marked tasks being marked and vice versa

kxrt committed:
Merge branch 'branch-A-BetterGui'

kxrt committed:
Add command to empty task list

kxrt committed:
Create _config.yml

kxrt committed:
Add Ui screenshot

kxrt committed:
Delete _config.yml

kxrt committed:
Update Ui screenshots

kxrt committed:
Create user guide, update empty command to reset

kxrt committed:
Fix User Guide image display issues

kxrt committed:
Update user guide styling

kxrt committed:
Update image sizes for User Guide

kxrt committed:
Change minor alignment formatting in User Guide

kxrt committed:
Edit alignment in User Guide

kxrt committed:
Align User Guide header to center

kxrt committed:
Update quick start instructions in User Guide

kxrt committed:
Update user guide and help command to include extension commands

kxrt committed:
Replace missing screenshot

kxrt committed:
Update user guide formatting

kxrt committed:
Fix background color on resize, rename storage file

kxrt committed:
Update User Guide
